singletons-th discards strictness in data type fields

Consider this example:

```hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE StandaloneKindSignatures #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE UndecidableInstances #-}
{-# OPTIONS_GHC -Wall #-}
module Bug where

import Data.Kind
import Data.Singletons.TH

$(singletons [d|
  type T :: Type -> Type
  data T a where
    MkT1 :: a -> T a
    MkT2 :: !Void -> T a
  |])

f :: T a -> a
f (MkT1 x) = x

type F :: T a -> a
type family F x where
  F ('MkT1 x) = x

sF :: forall a (x :: T a).
      Sing x -> Sing (F x)
sF (SMkT1 sx) = sx
```

When compiled, `f` will emit no warnings. But `sF`, its singled counterpart, will emit a warning:

```
Bug.hs:32:1: warning: [-Wincomplete-patterns]
    Pattern match(es) are non-exhaustive
    In an equation for ‘sF’: Patterns not matched: SMkT2 _
   |
32 | sF (SMkT1 sx) = sx
   | ^^^^^^^^^^^^^^^^^^
```

Why the discrepancy? It's because of one key different between the declarations for `T` and its singled counterpart, `ST`. The `MkT2` constructor of `T` declares its field of type `Void` as strict, which convinces GHC's pattern-match coverage checker that one cannot match on `MkT2` in `f`. On the hand, `ST` is declared like so:

```hs
type ST :: forall (a :: Type). T (a :: Type) -> Type
data ST z
  where
    SMkT1 :: forall a (n :: a). (Sing n) -> ST (MkT1 n :: T a)
    SMkT2 :: forall a (n :: Void). (Sing n) -> ST (MkT2 n :: T a)
```

This does _not_ declare `SMkT2`'s field as strict, which means that `sF` really is non-exhaustive, since one could invoke `sF (SMkT2 ⊥)`. I don't see any reason we shouldn't just carry strictness annotation from `MkT2` over to `SMkT2`, which would fix this issue.

--- a/README.md
+++ b/README.md
@@ -1006,6 +1006,21 @@ instance C [] where
   method = []
 ```
 
+## Support for singling, but not promotion
+
+The following constructs are supported for singleton generation but not promotion:
+
+* bang patterns
+
+See the following sections for more details.
+
+### Bang patterns
+
+Bang patterns (e.g., `f !x = (x, x)`) cannot be translated to a type-level
+setting as type families lack an equivalent of bang patterns. As a result,
+`singletons-th` will ignore any bang patterns and will simply promote the
+underyling pattern instead.
+
 ## Little to no support
 
 The following constructs are either unsupported or almost never work:
@@ -1015,6 +1030,8 @@ The following constructs are either unsupported or almost never work:
 * rank-n types
 * promoting `TypeRep`s
 * `TypeApplications`
+* Irrefutable patterns
+* `{-# UNPACK #-}` pragmas
 
 See the following sections for more details.
 
@@ -1322,3 +1339,25 @@ for the following constructs:
   type MSym0 :: forall a b. a ~> b ~> a
   type MSym1 :: forall a b. a -> b ~> a
   ```
+
+### Irrefutable patterns
+
+`singletons-th` will ignore irrefutable patterns (e.g., `f ~(x, y) = (y, x)`)
+and will simply promote or single the underlying patterns instead.
+`singletons-th` cannot promote irrefutable patterns for the same reason it
+cannot promote bang patterns: there is no equivalent syntax for type families.
+Moreover, `singletons-th` cannot single irrefutable patterns since singled data
+constructors are implemented as GADTs, as irrefutably matching on a GADT
+constructor will not bring the underlying type information into scope. Since
+essentially all singled code relies on using GADT type information in this way,
+it cannot reasonably be combined with irrefutable patterns, which prevent this
+key feature of GADT pattern matching.
+
+### `{-# UNPACK #-}` pragmas
+
+`singletons-th` will ignore `{-# UNPACK #-}` pragmas on the fields of a data
+constructor (e.g., `data T = MkT {-# UNPACK #-} !()`). This is because
+singled data types represent their argument types using existential type
+variables, and any data constructor that explicitly uses existential
+quantification cannot be unpacked. See
+[GHC#10016](https://gitlab.haskell.org/ghc/ghc/-/issues/10016).

--- a/singletons-base/tests/SingletonsBaseTestSuite.hs
+++ b/singletons-base/tests/SingletonsBaseTestSuite.hs
@@ -137,6 +137,7 @@ tests =
     , compileAndDumpStdTest "T450"
     , compileAndDumpStdTest "T453"
     , compileAndDumpStdTest "NegativeLiterals"
+    , compileAndDumpStdTest "T470"
     ],
     testCompileAndDumpGroup "Promote"
     [ compileAndDumpStdTest "Constructors"

--- a/singletons-base/tests/compile-and-dump/Singletons/T470.golden
+++ b/singletons-base/tests/compile-and-dump/Singletons/T470.golden
@@ -1,0 +1,92 @@
+Singletons/T470.hs:(0,0)-(0,0): Splicing declarations
+    singletons
+      [d| type T :: Type -> Type
+          
+          data T a
+            where
+              MkT1 :: a -> T a
+              MkT2 :: !Void -> T a
+          data S = MkS {-# UNPACK #-} !Bool |]
+  ======>
+    type T :: Type -> Type
+    data T a
+      where
+        MkT1 :: a -> T a
+        MkT2 :: !Void -> T a
+    data S = MkS {-# UNPACK #-} !Bool
+    type MkT1Sym0 :: (~>) a (T a)
+    data MkT1Sym0 :: (~>) a (T a)
+      where
+        MkT1Sym0KindInference :: SameKind (Apply MkT1Sym0 arg) (MkT1Sym1 arg) =>
+                                 MkT1Sym0 a0123456789876543210
+    type instance Apply MkT1Sym0 a0123456789876543210 = MkT1 a0123456789876543210
+    instance SuppressUnusedWarnings MkT1Sym0 where
+      suppressUnusedWarnings = snd (((,) MkT1Sym0KindInference) ())
+    type MkT1Sym1 :: a -> T a
+    type family MkT1Sym1 (a0123456789876543210 :: a) :: T a where
+      MkT1Sym1 a0123456789876543210 = MkT1 a0123456789876543210
+    type MkT2Sym0 :: (~>) Void (T a)
+    data MkT2Sym0 :: (~>) Void (T a)
+      where
+        MkT2Sym0KindInference :: SameKind (Apply MkT2Sym0 arg) (MkT2Sym1 arg) =>
+                                 MkT2Sym0 a0123456789876543210
+    type instance Apply MkT2Sym0 a0123456789876543210 = MkT2 a0123456789876543210
+    instance SuppressUnusedWarnings MkT2Sym0 where
+      suppressUnusedWarnings = snd (((,) MkT2Sym0KindInference) ())
+    type MkT2Sym1 :: Void -> T a
+    type family MkT2Sym1 (a0123456789876543210 :: Void) :: T a where
+      MkT2Sym1 a0123456789876543210 = MkT2 a0123456789876543210
+    type MkSSym0 :: (~>) Bool S
+    data MkSSym0 :: (~>) Bool S
+      where
+        MkSSym0KindInference :: SameKind (Apply MkSSym0 arg) (MkSSym1 arg) =>
+                                MkSSym0 a0123456789876543210
+    type instance Apply MkSSym0 a0123456789876543210 = MkS a0123456789876543210
+    instance SuppressUnusedWarnings MkSSym0 where
+      suppressUnusedWarnings = snd (((,) MkSSym0KindInference) ())
+    type MkSSym1 :: Bool -> S
+    type family MkSSym1 (a0123456789876543210 :: Bool) :: S where
+      MkSSym1 a0123456789876543210 = MkS a0123456789876543210
+    type ST :: forall (a :: Type). T a -> Type
+    data ST :: forall (a :: Type). T a -> Type
+      where
+        SMkT1 :: forall a (n :: a). (Sing n) -> ST (MkT1 n :: T a)
+        SMkT2 :: forall a (n :: Void). !(Sing n) -> ST (MkT2 n :: T a)
+    type instance Sing @(T a) = ST
+    instance SingKind a => SingKind (T a) where
+      type Demote (T a) = T (Demote a)
+      fromSing (SMkT1 b) = MkT1 (fromSing b)
+      fromSing (SMkT2 b) = MkT2 (fromSing b)
+      toSing (MkT1 (b :: Demote a))
+        = case toSing b :: SomeSing a of {
+            SomeSing c -> SomeSing (SMkT1 c) }
+      toSing (MkT2 (b :: Demote Void))
+        = case toSing b :: SomeSing Void of {
+            SomeSing c -> SomeSing (SMkT2 c) }
+    data SS :: S -> Type
+      where SMkS :: forall (n :: Bool). !(Sing n) -> SS (MkS n :: S)
+    type instance Sing @S = SS
+    instance SingKind S where
+      type Demote S = S
+      fromSing (SMkS b) = MkS (fromSing b)
+      toSing (MkS (b :: Demote Bool))
+        = case toSing b :: SomeSing Bool of {
+            SomeSing c -> SomeSing (SMkS c) }
+    instance SingI n => SingI (MkT1 (n :: a)) where
+      sing = SMkT1 sing
+    instance SingI (MkT1Sym0 :: (~>) a (T a)) where
+      sing = (singFun1 @MkT1Sym0) SMkT1
+    instance SingI n => SingI (MkT2 (n :: Void)) where
+      sing = SMkT2 sing
+    instance SingI (MkT2Sym0 :: (~>) Void (T a)) where
+      sing = (singFun1 @MkT2Sym0) SMkT2
+    instance SingI n => SingI (MkS (n :: Bool)) where
+      sing = SMkS sing
+    instance SingI (MkSSym0 :: (~>) Bool S) where
+      sing = (singFun1 @MkSSym0) SMkS
+
+Singletons/T470.hs:0:0: warning:
+    {-# UNPACK #-} pragmas are ignored by `singletons-th`.
+  |
+6 | $(singletons [d|
+  |   ^^^^^^^^^^^^^^...

--- a/singletons-base/tests/compile-and-dump/Singletons/T470.hs
+++ b/singletons-base/tests/compile-and-dump/Singletons/T470.hs
@@ -1,0 +1,24 @@
+module T470 where
+
+import Data.Kind
+import Data.Singletons.Base.TH
+
+$(singletons [d|
+  type T :: Type -> Type
+  data T a where
+    MkT1 :: a -> T a
+    MkT2 :: !Void -> T a
+
+  data S = MkS {-# UNPACK #-} !Bool
+  |])
+
+f :: T a -> a
+f (MkT1 x) = x
+
+type F :: T a -> a
+type family F x where
+  F ('MkT1 x) = x
+
+sF :: forall a (x :: T a).
+      Sing x -> Sing (F x)
+sF (SMkT1 sx) = sx

--- a/singletons-th/CHANGES.md
+++ b/singletons-th/CHANGES.md
@@ -102,6 +102,8 @@ Changelog for singletons-th project
   default implementation. Use `singEqInstance{s}` instead.
 * The Template Haskell machinery will no longer promote `TypeRep` to `Type`,
   as this special case never worked properly in the first place.
+* The Template Haskell machinery will now preserve strict fields in data types
+  when generating their singled counterparts.
 * Introduce a new `promotedDataTypeOrConName` option to
   `Data.Singletons.TH.Options`. Overriding this option can be useful in
   situations where one wishes to promote types such as `Nat`, `Symbol`, or

--- a/singletons-th/src/Data/Singletons/TH/Single/Data.hs
+++ b/singletons-th/src/Data/Singletons/TH/Single/Data.hs
@@ -196,11 +196,13 @@ singCtor dataName (DCon con_tvbs cxt name fields rty)
   -- just assume that the context is always ().
   emitDecs =<< singDefuns name DataName [] (map Just kinds) (Just rty')
 
-  let noBang    = Bang NoSourceUnpackedness NoSourceStrictness
-      args      = map ((noBang,) . DAppT singFamily) indices
-      conFields = case fields of
-                    DNormalC dInfix _ -> DNormalC dInfix args
-                    DRecC _           -> DNormalC False  args
+  conFields <- case fields of
+    DNormalC dInfix bts -> DNormalC dInfix <$>
+                           zipWithM (\(b, _) index -> mk_bang_type b index)
+                                    bts indices
+    DRecC vbts          -> DNormalC False <$>
+                           zipWithM (\(_, b, _) index -> mk_bang_type b index)
+                                    vbts indices
                       -- Don't bother looking at record selectors, as they are
                       -- handled separately in singTopLevelDecs.
                       -- See Note [singletons-th and record selectors]
@@ -210,6 +212,24 @@ singCtor dataName (DCon con_tvbs cxt name fields rty)
                   -- Make sure to include an explicit `rty'` kind annotation.
                   -- See Note [Preserve the order of type variables during singling],
                   -- wrinkle 3, in D.S.TH.Single.Type.
+  where
+    mk_source_unpackedness :: SourceUnpackedness -> SgM SourceUnpackedness
+    mk_source_unpackedness su = case su of
+      NoSourceUnpackedness -> pure su
+      SourceNoUnpack       -> pure su
+      SourceUnpack         -> do
+        -- {-# UNPACK #-} is essentially useless in a singletons setting, since
+        -- all singled data types are GADTs. See GHC#10016.
+        qReportWarning "{-# UNPACK #-} pragmas are ignored by `singletons-th`."
+        pure NoSourceUnpackedness
+
+    mk_bang :: Bang -> SgM Bang
+    mk_bang (Bang su ss) = do su' <- mk_source_unpackedness su
+                              pure $ Bang su' ss
+
+    mk_bang_type :: Bang -> DType -> SgM DBangType
+    mk_bang_type b index = do b' <- mk_bang b
+                              pure (b', DAppT singFamily index)
 
 {-
 Note [singletons-th and record selectors]

